clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1053,6 +1060,52 @@ static target getAccessTarget(QualType FieldTy,
       AccTy->getTemplateArgs()[3].getAsIntegral().getExtValue());
 }
 
+static bool IsFreeFunction(Sema &SemaRef, const FunctionDecl *FD) {
+  if (FD->hasAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    const auto &Attrs = FD->getAttrs();
+    for (unsigned i = 0, e = Attrs.size(); i < e; i++) {
+      const SYCLAddIRAttributesFunctionAttr *AnAttr =
+          dyn_cast<SYCLAddIRAttributesFunctionAttr>(Attrs[i]);
+      if (AnAttr) {
+        SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
+            AnAttr->getAttributeNameValuePairs(SemaRef.Context);
+        for (const auto &NameValuePair : NameValuePairs) {
+          if (NameValuePair.first == "sycl-nd-range-kernel" ||
+              NameValuePair.first == "sycl-single-task-kernel")
+            return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+static std::string constructFFKernelName(const FunctionDecl *FD) {
+  IdentifierInfo *Id = FD->getIdentifier();
+  std::string NewIdent = (Twine("__free_function_") + Id->getName()).str();
+  return NewIdent;
+}
+
+// Gets a name for the free function kernel function. The suffix allows a normal
+// device function to coexist with the kernel function.
+static std::pair<std::string, std::string>
+constructFreeFunctionKernelName(Sema &S, const FunctionDecl *FreeFunc,
+                                MangleContext &MC) {
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+  std::string MangledName;
+  std::string StableName;
+
+  if (FreeFunc->getTemplateSpecializationArgs()) {
+    MC.mangleName(FreeFunc, Out);
+    MangledName = (Twine("__free_function") + Out.str()).str();
+  } else {
+    MangledName = constructFFKernelName(FreeFunc);
+  }
```

Comment:
For this and your previous comment about naming of the free functions, can you help with doing what you suggest: creating a name mangling extension. Is there an example of that I can adapt?

Reply:
Yes, I can help with that. I recommend we use the mangled name of a function template specialization with a reserved name (e.g., `__sycl_kernel`) that has a single non-type template parameter corresponding to the function name. See https://godbolt.org/z/qG16jasnr where the `__sycl_kernel` specialization generated for function `f` is given the following mangled names:

- Itanium: `_Z13__sycl_kernelITnDaXadL_Z1fiEEEvv`
- Microsoft: `??$__sycl_kernel@$MP6AXH@Z1?f@@YAXH@Z@@YAXXZ`

Unfortunately, it seems the `llvm-cxxfilt` and `llvm-undname` utilities are missing support for some of the terms used in those manglings and neither is able to unmangle them at present. Fixing those should be pretty straight forward and should produce results similar to `void __sycl_kernel<&f(int)>()`.

(it looks like the `Tn` term in the Itanium mangled name is unknown to the demangler; it also isn't specified in the [Itanium ABI](http://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling). I'm guessing the situation is similar for the Microsoft mangling, but there is no specification to refer to for it).

I see the added tests are passing the mangled name of the kernel to `get_kernel()`. The free function extension doc doesn't specify a function that does the lookup by mangled name, so I assume this is just for testing. We certainly wouldn't want to expose these mangled names directly to users.

Reply:
Yes, the mechanism to get a kernel_id in order to launch the kernel is a stop-gap. Once the get_kernel_id and other related APIs are implemented, the test will be changed and the invocation mechanism will conform to the free function spec.

Reply:
@tahonermann, are you going to supply some actual code that does the proposed naming you have described, or are you leaving it to me to try and do it?

Reply:
Looping @cperkinsintel into this conversation.  Remember that the [kernel compiler](https://github.com/intel/llvm/pull/11985) needs to be able to lookup the kernel function by its (mangled) name.  If the kernel is defined `extern "C"` like:

```
extern "C"
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclexp::nd_range_kernel<1>))
void foo(int *in, int *out) {/*...*/}
```

then the application can query the kernel using the same name that appears in the source code like:

```
sycl::kernel k = kb.ext_oneapi_get_kernel("foo");
```

If the front-end is really mangling the name as @tahonermann proposes, the SYCL runtime will somehow need to know how to map "foo" to that mangled name.  This might be simpler if the mangling simply added a prefix like "__sycl_kernel_foo".  Is there a reason we don't do something simple like this?

We will also need to figure out how the SYCL runtime implements the query when function is not `extern "C"`.  In this case, the application needs to "register" the unmangled name of the function with the kernel compiler as described [here](https://github.com/intel/llvm/blob/13581a3a57445fab96d1d1aa567ee6872f4ea1bb/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#using-the-registered_kernel_names-property).  I think this probably implies that the compiler somehow passes some metadata to the SYCL runtime with the mapping.  For example, maybe this can be done using our property set infrastructure.

Reply:
@rdeodhar, I'm unlikely to have time to contribute code in the time frame needed for this project. But I can help to identify functions that will need to be called and answer any questions about the mangling.

@gmlueck , I wasn't aware of the kernel compiler requirement. If I'm understanding things correctly, these names will have more persistence than I expected and changing them later could be complicated. That is a little concerning.

For functions declared `extern "C"`, I agree with using a simple prefix; `__sycl_kernel_<xxx>` sounds good to me.

Correct me if I'm mistaken, but it will be possible for multiple functions with the same name (in the same declaration context) to be used as free function kernels, yes? If so, then either some form of mangling is required or the free functions have to be given an explicit name in some other way as can be done for regular SYCL kernels. Regardless, I think we should produce a symbol that can be demangled by existing demanglers.

I've been looking at the linked doc (`sycl_ext_oneapi_kernel_compiler.asciidoc`). There are things I don't understand. In those examples, how is `source` associated with `kb` or `kb_src`? There seems to be something missing there.

Reply:
Would it help if the free function kernel name were constructed by prefixing the function id with __free_function_ (or __sycl_kernel_) and then mangled, instead of adding the prefix to the already-mangled name? Since all the free functions have unique mangled names before adding the prefix, the generated names will also be unique, and could be demangled by standard utilities.

Reply:
I also think that @rdeodhar's proposal makes sense.  This also unifies the naming scheme with `extern "C"` kernels.  The algorithm would be: prefix the unmangled name with "\_\_sycl_kernel\_" and then mangle the name if normal C++ rules would require mangling.

> I've been looking at the linked doc (`sycl_ext_oneapi_kernel_compiler.asciidoc`). There are things I don't understand. In those examples, how is `source` associated with `kb` or `kb_src`? There seems to be something missing there.

For a complete example that shows the connections, see [Simple example](https://github.com/intel/llvm/blob/13581a3a57445fab96d1d1aa567ee6872f4ea1bb/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#simple-example) later in that document.  For the complete details of the extension, you will need to read from the top of the document.

> I wasn't aware of the kernel compiler requirement. If I'm understanding things correctly, these names will have more persistence than I expected and changing them later could be complicated. That is a little concerning.

I was imagining that the front-end will pass metadata with information that maps the unmangled name to the mangled name, and the SYCL runtime will use this mapping information.  Therefore, the SYCL runtime won't duplicate the name mangling algorithm, so we could change that algorithm in the front-end later if we wanted.  To illustrate, consider the example [here](https://github.com/intel/llvm/blob/13581a3a57445fab96d1d1aa567ee6872f4ea1bb/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc#instantiating-templated-kernel-functions).  The user application register two instantiations of the kernel function template:

```
sycl::kernel k_float = kb.ext_oneapi_get_kernel("bartmpl<float>");
sycl::kernel k_int = kb.ext_oneapi_get_kernel("bartmpl<int>");
```

The metadata could contain two entries:

```
bartmpl<float>: __mangled name for this kernel__
bartmpl<int>:  __mangled name for this kernel__
```

The application passes either the string "bartmpl\<float\>" or "bartmpl\<int\>" to the runtime.  The runtime uses the table to find the associated mangled name, and this mangled name is passed to `zeKernelCreate` to create the Level Zero kernel object.

Reply:
I've changed the free function kernel name so it is demangle-able and is unique because it has __sycl_kernel_ prepended to the original free function name.

Reply:
The function constructFreeFunctionKernelName creates a FunctionDecl solely for the purpose of being able to generate a mangled name. Can that FunctionDecl be left as is or should it be deleted before exiting the function using "delete"?